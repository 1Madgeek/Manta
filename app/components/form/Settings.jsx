// Libraries
import React from 'react';
import PropTypes from 'prop-types';

// Animation
import {Motion, spring} from 'react-motion';

// Styles
import styled from 'styled-components';

const Wrapper = styled.div`
  position: relative;
  margin-bottom: 20px;
  overflow: hidden;
  border-radius: 4px 4px 0 0;
  border-bottom: 1px solid rgba(0,0,0,.1);
`;

const SettingsHeader = styled.a`
  border-bottom: 1px solid rgba(0,0,0,.1);
  padding: 10px 40px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  text-decoration: none;
  color: #2C323A;
  &:hover {
    text-decoration: none;
    color: #2C323A;
  }
`;

const AllSettings = styled.div`
  display: flex;
  padding: 25px 40px;
  background: #F2F3F4;
`;

const Setting = styled.div`
  margin-right: 20px;
  > label {
    color: #4F555C;
    margin-bottom: 10px;
 }
`;

const Label = styled.label`
  font-weight: 600;
  font-size: 12px;
  text-transform: uppercase;
  letter-spacing: 1px;
  color: #4F555C;
  margin-bottom: 0px;
`;

// Custom Components
import Switch from '../shared/Switch';

// Component
function Settings(props) {
  const {
    isSettingsOpened,
    toggleFormSettings,
    toggleField,
    currentInvoice,
  } = props;

  const {
    dueDate,
    currency,
    discount,
    vat,
    note,
  } = currentInvoice;

  return (
    <Motion
      style={{
        height: spring(isSettingsOpened ? 155 : 45),
        rotate: spring(isSettingsOpened ? 180 : 0),
      }}>
      {({height, rotate}) =>
        <Wrapper style={{ height: `${height}px`}}>
          <SettingsHeader
            href="#"
            onClick={toggleFormSettings}>
            <Label>Form Settings</Label>
            <div style={{
                transform: `rotate(${rotate}deg)`
              }}>
              <i className="ion-arrow-down-b"/>
            </div>
          </SettingsHeader>

          <AllSettings>
            <Setting>
              <Label>Due Date</Label>
              <Switch
                checked={dueDate.required}
                onChange={() => toggleField('dueDate')}
              />
            </Setting>

            <Setting>
              <Label>Currency</Label>
              <Switch
                checked={currency.required}
                onChange={() => toggleField('currency')}
              />
            </Setting>

            <Setting>
              <Label>Discount</Label>
              <Switch
                checked={discount.required}
                onChange={() => toggleField('discount')}
              />
            </Setting>

            <Setting>
              <Label>Vat</Label>
              <Switch
                checked={vat.required}
                onChange={() => toggleField('vat')}
              />
            </Setting>

            <Setting>
              <Label>Note</Label>
              <Switch
                checked={note.required}
                onChange={() => toggleField('note')}
              />
            </Setting>
          </AllSettings>
        </Wrapper>}
    </Motion>
  );
}

Settings.propTypes = {
  isSettingsOpened: PropTypes.bool.isRequired,
  currentInvoice: PropTypes.object.isRequired,
  toggleField: PropTypes.func.isRequired,
  toggleFormSettings: PropTypes.func.isRequired,
};

// Export
export default Settings;
