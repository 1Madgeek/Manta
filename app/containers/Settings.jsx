// React
import React, {Component} from 'react';
import PropTypes from 'prop-types';

// Redux
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as ActionCreators from '../actions/settings';

// 3rd Party Libs
const _ = require('lodash');

// Layout
import {
  PageWrapper,
  PageHeader,
  PageHeaderTitle,
  PageHeaderActions,
  PageContent,
  PageFooter,
  } from '../components/shared/Layout';

// Animation
import _withFadeInAnimation from '../components/shared/hoc/_withFadeInAnimation';

// Components
import Info from '../components/settings/Info';
import AppSettings from '../components/settings/AppSettings';
import PrintOptions from '../components/settings/PrintOptions';
import Message from '../components/shared/Message';
import Button from '../components/shared/Button';
import {
  TabsWrapper,
  TabContent,
  Tab,
  } from '../components/shared/Tabs';

// Component
class Settings extends Component {
  componentWillMount = () => {
    this.setState({  visibleTab: 1  });
  }

  // Check if settings have been saved
  settingsSaved = () => {
    const {current, saved} = this.props.settings;
    return _.isEqual(current, saved);
  };

  // Save Settings to App Config
  saveSettingsState = () => {
    // Dispatch Action
    const {dispatch} = this.props;
    const saveSettings = bindActionCreators(
      ActionCreators.saveSettings,
      dispatch,
    );
    saveSettings(this.props.settings.current);
  };

  // Update Info Settings
  updateInfo = data => {
    const {dispatch} = this.props;
    const updateInfo = bindActionCreators(ActionCreators.updateInfo, dispatch);
    updateInfo(data);
  };

  // Update App Settings
  updateAppSettings = data => {
    const {dispatch} = this.props;
    const updateAppSettings = bindActionCreators(
      ActionCreators.updateAppSettings,
      dispatch,
    );
    updateAppSettings(data);
  };

  // Update Print Options
  updatePrintOptions = data => {
    const {dispatch} = this.props;
    const updatePrintOptions = bindActionCreators(
      ActionCreators.updatePrintOptions,
      dispatch,
    );
    updatePrintOptions(data);
  }

  // Switch Tab
  changeTab = tabNum => {
    this.setState({visibleTab: tabNum});
  };

  render = () => {
    const {info, appSettings, printOptions} = this.props.settings.current;
    return (
      <PageWrapper>
        <PageHeader>
          <PageHeaderTitle>Settings</PageHeaderTitle>
          {!this.settingsSaved() &&
            <PageHeaderActions>
              <Button primary onClick={() => this.saveSettingsState()}>
                Save
              </Button>
          </PageHeaderActions>}
        </PageHeader>
        <PageContent>
          <TabsWrapper>
            <Tab
              href="#"
              className={this.state.visibleTab === 1 ? 'active' : ''}
              onClick={() => this.changeTab(1)}>
              Profile
            </Tab>
            <Tab
              href="#"
              className={this.state.visibleTab === 2 ? 'active' : ''}
              onClick={() => this.changeTab(2)}>
              Print Options
            </Tab>
            <Tab
              href="#"
              className={this.state.visibleTab === 3 ? 'active' : ''}
              onClick={() => this.changeTab(3)}>
              App Settings
            </Tab>
          </TabsWrapper>
          <TabContent>
            {this.state.visibleTab === 1 &&
              <Info info={info} updateInfo={this.updateInfo} />}
            {this.state.visibleTab === 2 &&
                <PrintOptions
                  printOptions={printOptions}
                  updatePrintOptions={this.updatePrintOptions} />}
            {this.state.visibleTab === 3 &&
              <AppSettings
                appSettings={appSettings}
                updateAppSettings={this.updateAppSettings}
              />}
          </TabContent>
        </PageContent>
      </PageWrapper>
    );
  };
}

// PropTypes Validation
Settings.propTypes = {
  settings: PropTypes.object.isRequired,
};

// Map state to props & Added Faded In Animation
Settings = connect(state => ({ settings: state.SettingsReducer }))(Settings);
Settings = _withFadeInAnimation(Settings);

// Export
export default Settings;
