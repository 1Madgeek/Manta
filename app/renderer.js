// Node Libs
const path = require('path');
const glob = require('glob');

// Custom Libs
const sounds = require(path.join(__dirname, '../libs/sounds.js'));

// Imports Main Renderere Files
function loadMaiRendererFiles() {
  const files = glob.sync(path.join(__dirname, './renderers/*.js'));
  files.forEach(file => require(file));
  sounds.preload();
}

function initialize() {
  loadMaiRendererFiles();
  sounds.play('STARTUP');
  console.timeEnd('init');
}

initialize();
