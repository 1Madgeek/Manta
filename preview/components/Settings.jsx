// Libs
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Styling
import styled from 'styled-components';

const Wrapper = styled.div`
`;

const Label = styled.label`
  font-size: 12px;
  text-transform: uppercase;
  letter-spacing: 1px;
  font-weight: 600;
  color: #4F555C;
`;

const Section = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 30px;
  justify-content: flex-end;
  align-items: flex-start;
`;

const Select = styled.select`
  width: 150px;
  height: 24px;
  margin-right: 20px;
`;

class Settings extends Component {
  render() {
    const {settings, updateSettings} = this.props;
    return (
      <Wrapper>
        <Section>
          <Label>Template</Label>
          <Select
            name="template"
            value={settings.template}
            onChange={updateSettings}>
            <option value="default">Default</option>
            <option value="hosting">Hosting</option>
            <option value="elegant">Elegant</option>
            <option value="classic">Classic</option>
          </Select>
        </Section>

        <Section>
          <Label>Margin Type</Label>
          <Select
            name="marginsType"
            value={settings.marginsType}
            onChange={updateSettings}>
            <option value="0">Default Margin</option>
            <option value="1">No Margin</option>
            <option value="2">Minimum Margin</option>
          </Select>
        </Section>

        <Section>
          <Label className="itemLabel">Page Size</Label>
          <Select
            name="pageSize"
            value={settings.pageSize}
            onChange={updateSettings}>
            <option value="A3">A3</option>
            <option value="A4">A4</option>
            <option value="A5">A5</option>
            <option value="Legal">Legal</option>
            <option value="Letter">Letter</option>
            <option value="Tabloid">Tabloid</option>
          </Select>
        </Section>
      </Wrapper>
    );
  }
}

Settings.propTypes = {
  settings: PropTypes.object.isRequired,
  updateSettings: PropTypes.func.isRequired,
};

export default Settings;
