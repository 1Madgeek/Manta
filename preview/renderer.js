// Electron Libs
const ipc = require('electron').ipcRenderer;

// Libraries
import React from 'react';
import ReactDOM from 'react-dom';

// Custom Components
import Receipt from './containers/Receipt.jsx';

// Render Preview Window
ipc.on('update-preview', (event, receiptData) => {
  ReactDOM.render(
    <Receipt receiptData={receiptData} />,
    document.getElementById('receiptWrapper')
  );
});

