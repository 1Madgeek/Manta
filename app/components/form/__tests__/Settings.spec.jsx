// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import Switch from '../../shared/Switch';

// Component
import Settings from '../Settings.jsx';

// Mocks
const currentInvoice = {
  recipient: {
    newRecipient: true,
    new: {},
    select: {},
  },
  rows: [],
  dueDate: { required: false },
  currency: { required: false },
  discount: { required: false },
  tax: { required: false },
  note: { required: false },
  settingsOpen: false,
};
const toggleField = jest.fn();
const toggleFormSettings = jest.fn();

describe('Settings component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Settings
        toggleField={toggleField}
        toggleFormSettings={toggleFormSettings}
        currentInvoice={currentInvoice}
      />,
    );
  });

  // PROPS & STATE
  it('receives correct props', () => {
    expect(wrapper.prop('toggleField')).toEqual(toggleField);
    expect(wrapper.prop('toggleFormSettings')).toEqual(toggleFormSettings);
    expect(wrapper.prop('currentInvoice')).toEqual(currentInvoice);
  });

  // RENDER
  it('renders necessary element', () => {
    expect(wrapper.find('label')).toHaveLength(11);
    expect(wrapper.find(Switch)).toHaveLength(5);
  });

  // LIFE CYCLE EVENTS
  // TODO
  it('render when necessary', () => {});

  // PRIsettingsE METHOD
  it('toggle field correctly', () => {
    // Setup
    const dueDate  = wrapper.find(Switch).at(0);
    const currency = wrapper.find(Switch).at(1);
    const discount = wrapper.find(Switch).at(2);
    const tax      = wrapper.find(Switch).at(3);
    const note     = wrapper.find(Switch).at(4);

    // Execute & Assert
    dueDate.find('input').simulate('change');
    expect(toggleField).toHaveBeenCalled();
    expect(toggleField).toHaveBeenCalledWith('dueDate');
    expect(toggleField).not.toHaveBeenCalledWith('something-else');

    currency.find('input').simulate('change');
    expect(toggleField).toHaveBeenCalled();
    expect(toggleField).toHaveBeenCalledWith('currency');
    expect(toggleField).not.toHaveBeenCalledWith('something-else');

    discount.find('input').simulate('change');
    expect(toggleField).toHaveBeenCalled();
    expect(toggleField).toHaveBeenCalledWith('discount');
    expect(toggleField).not.toHaveBeenCalledWith('something-else');

    tax.find('input').simulate('change');
    expect(toggleField).toHaveBeenCalled();
    expect(toggleField).toHaveBeenCalledWith('tax');
    expect(toggleField).not.toHaveBeenCalledWith('something-else');

    note.find('input').simulate('change');
    expect(toggleField).toHaveBeenCalled();
    expect(toggleField).toHaveBeenCalledWith('note');
    expect(toggleField).not.toHaveBeenCalledWith('something-else');
  });

  // SNAPSHOT
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Settings
          toggleField={toggleField}
          toggleFormSettings={toggleFormSettings}
          currentInvoice={currentInvoice}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
