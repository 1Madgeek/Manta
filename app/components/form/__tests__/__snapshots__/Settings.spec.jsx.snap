// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component matches snapshot 1`] = `
<div
  className="Settings__Wrapper-iXFqbq jshSii"
  style={
    Object {
      "height": "45px",
    }
  }
>
  <a
    className="Settings__SettingsHeader-hbrDNs hKQuKk"
    href="#"
    onClick={[MockFunction]}
  >
    <label
      className="Settings__Label-bIhYmz iPSgUQ"
    >
      Form Settings
    </label>
    <div
      style={
        Object {
          "transform": "rotate(0deg)",
        }
      }
    >
      <i
        className="ion-arrow-down-b"
      />
    </div>
  </a>
  <div
    className="Settings__Helper-cagYXW jnzDbx"
  >
    <p>
      Toogle any field to make it required in the form.
    </p>
  </div>
  <div
    className="Settings__AllSettings-hAEgKd cLFeXF"
  >
    <div
      className="Settings__Setting-iSwZaF jiKDp"
    >
      <label
        className="Settings__Label-bIhYmz iPSgUQ"
      >
        Due Date
      </label>
      <label
        className="Switch__SwitchStyle-cIZUZy lcgmHp"
      >
        <input
          checked={false}
          className="Switch__Input-jIJATX kGoGgc"
          name="dueDate"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Switch__Slider-iwwTvI egztuK"
        />
      </label>
    </div>
    <div
      className="Settings__Setting-iSwZaF jiKDp"
    >
      <label
        className="Settings__Label-bIhYmz iPSgUQ"
      >
        Currency
      </label>
      <label
        className="Switch__SwitchStyle-cIZUZy lcgmHp"
      >
        <input
          checked={false}
          className="Switch__Input-jIJATX kGoGgc"
          name="currency"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Switch__Slider-iwwTvI egztuK"
        />
      </label>
    </div>
    <div
      className="Settings__Setting-iSwZaF jiKDp"
    >
      <label
        className="Settings__Label-bIhYmz iPSgUQ"
      >
        Discount
      </label>
      <label
        className="Switch__SwitchStyle-cIZUZy lcgmHp"
      >
        <input
          checked={false}
          className="Switch__Input-jIJATX kGoGgc"
          name="discount"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Switch__Slider-iwwTvI egztuK"
        />
      </label>
    </div>
    <div
      className="Settings__Setting-iSwZaF jiKDp"
    >
      <label
        className="Settings__Label-bIhYmz iPSgUQ"
      >
        Tax
      </label>
      <label
        className="Switch__SwitchStyle-cIZUZy lcgmHp"
      >
        <input
          checked={false}
          className="Switch__Input-jIJATX kGoGgc"
          name="tax"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Switch__Slider-iwwTvI egztuK"
        />
      </label>
    </div>
    <div
      className="Settings__Setting-iSwZaF jiKDp"
    >
      <label
        className="Settings__Label-bIhYmz iPSgUQ"
      >
        Note
      </label>
      <label
        className="Switch__SwitchStyle-cIZUZy lcgmHp"
      >
        <input
          checked={false}
          className="Switch__Input-jIJATX kGoGgc"
          name="note"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="Switch__Slider-iwwTvI egztuK"
        />
      </label>
    </div>
  </div>
</div>
`;
